### PR TITLE
[SPARK-50598][SQL] An initial, no-op PR which adds new parameters to already existing classes UnresolvedWith, CTERelationRef and CTERelationDef to enable later implementation of recursive CTEs.

### PR DESCRIPTION
**Pull request by milanisvet**
### What changes were proposed in this pull request?

PR adds new parameters to already existing classes UnresolvedWith, CTERelationRef and CTERelationDef to enable later implementation of recursive CTEs. 
- Additional parameters are added in pattern matching cases in other files of mentioned classes as well. 
- A large number of trivial changes in tests was required, due to new parameters introduced to the classes, which are also addressed in this PR. 
- More information for reviewers can be found here: https://docs.google.com/document/d/1qcEJxqoXcr5cSt6HgIQjWQSqhfkSaVYkoDHsg5oxXp4/edit 

### Why are the changes needed?

Support for the recursive CTE.

### Does this PR introduce _any_ user-facing change?

No. RECURSIVE keyword is not introduced in this PR.

### How was this patch tested?

The tests failing after the initial change were all in `SQLQueryTestSuite`, so this patch was tested by running this Test Suite.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by MaxGekk** [comment]
@peter-toth Please, have a look at the PR.

**Comment by milanisvet** [comment]
> @peter-toth Please, have a look at the PR.

We already contacted Peter and I don't think he will have time to review it. Could you have a look at it? This PR are only small no-op changes to enable later introduction to recursive CTEs.

**Comment by peter-toth** [comment]
I've just taken a quick look, and this PR seems to just extract some changes from my https://github.com/apache/spark/pull/40744 so I'm fine with these changes.

But @milanisvet , if you want to split the changes into smaller pieces then probably [SPARK-24497](https://issues.apache.org/jira/browse/SPARK-24497) should be converted to an umbrella/epic ticket and the changes should be subtasks of it.

@cloud-fan, @dongjoon-hyun what is the preffered way of organizing the work?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 1]
```diff
@@ -830,12 +832,19 @@ case class UnresolvedWith(
  *                                   pushdown to help ensure rule idempotency.
  * @param underSubquery If true, it means we don't need to add a shuffle for this CTE relation as
  *                      subquery reuse will be applied to reuse CTE relation output.
+ * @param recursive If true the definition is recursive.
+ * @param recursionAnchor A helper plan node that temporary stores the anchor term of recursive
+ *                        definitions. In the beginning of recursive resolution the `ResolveWithCTE`
+ *                        rule updates this parameter and once it is resolved the same rule resolves
+ *                        the recursive [[CTERelationRef]] references and removed this parameter.
```
[the remark this answers is not in the document]
Comment (milanisvet) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: Corrected. Thanks :)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 2]
```diff
@@ -830,12 +832,19 @@ case class UnresolvedWith(
  *                                   pushdown to help ensure rule idempotency.
  * @param underSubquery If true, it means we don't need to add a shuffle for this CTE relation as
  *                      subquery reuse will be applied to reuse CTE relation output.
+ * @param recursive If true the definition is recursive.
```
[the remark this answers is not in the document]
Comment (milanisvet) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: The definition is recursive meaning that it contains a self-reference. This is corrected now. Thanks :)

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by cloud-fan** [comment]
Oh, I'll merge after the proper JIRA ticket is created.

**Review by MaxGekk — APPROVED** [review]

**Comment by milanisvet** [comment]
JIRA subticket created, and the PR description is changed with the new JIRA ticket number @cloud-fan 
Thanks again for approving :)

**Pull request closed without merging**